[TIMOB-16878] Android: Refactored ScrollableView

jquick-axway opened the pull request:
**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-5996
- https://jira.appcelerator.org/browse/TIMOB-15780
- https://jira.appcelerator.org/browse/TIMOB-16878
- https://jira.appcelerator.org/browse/TIMOB-25539
- https://jira.appcelerator.org/browse/TIMOB-25625
- https://jira.appcelerator.org/browse/TIMOB-25626
- https://jira.appcelerator.org/browse/TIMOB-25634
- https://jira.appcelerator.org/browse/TIMOB-25636

**Summary:**
- [TIMOB-5996](https://jira.appcelerator.org/browse/TIMOB-5996) Modified `ScrollableView.removeView()` to support integer indexes like iOS.
- [TIMOB-15780](https://jira.appcelerator.org/browse/TIMOB-15780) Fixed `ScrollableView.setCurrentPage()` to not fire a "scrollend" event.
  * Event should only be fired when dragged or for methods that trigger scrolling animations.
  * Added missing "dragstart" event to be paired with existing "dragend" event. Now matches docs.
- [TIMOB-16878](https://jira.appcelerator.org/browse/TIMOB-16878) Fixed bug where ScrollableView ignores top-most child view's layout properties.
  * Prevented properties width, height, top, bottom, left, and right from working.
  * Fixed by wrapping view with a `TiCompositeLayout`, which is needed to handle these layout properties.
- [TIMOB-25539](https://jira.appcelerator.org/browse/TIMOB-25539) Fixed crash caused by adding a border to the ScrollableView.
  * Bug was introduced in Titanium 6.1.0.
- [TIMOB-25625](https://jira.appcelerator.org/browse/TIMOB-25625) Modified left/right page arrows to be density scaled.
  * Was hard-code to 80 pixels for all screens.
  * Now uses base mdpi (1x) resolution of 24 pixels and scales based on DPI from there.
- [TIMOB-25626](https://jira.appcelerator.org/browse/TIMOB-25626) Fixed ScrollableView to support `Ti.UI.SIZE` for width.
  * Android's `ViewPager` Java class does not natively support `WRAP_CONTENT`.
  * Refactored onMeasure() code to re-size width/height of container based on largest child view in this case.
- [TIMOB-25634](https://jira.appcelerator.org/browse/TIMOB-25634) Modified ScrollableView so that `cacheSize` property can be set dynamically like iOS.
  * Added methods `setCacheSize()` and `getCacheSize()`.
  * Can no longer be set less than 3, matching iOS' min cache size.
  * Now better converts iOS style cache size value to Android offscreen page count value.
- [TIMOB-25636](https://jira.appcelerator.org/browse/TIMOB-25636) Fixed bug where calling ScrollableView setters cause a crash if parent window is closed.
  * Bug was introduced as of Titanium 7.0.0.
- Removed API Level 9 guards from code.
  * Outdated code. Min API Level is currently 16.

---
**Test 1:**
This test ensures border applied to ScrollableView does not cause a crash.
1. Build and run the code attached to [TIMOB-25539](https://jira.appcelerator.org/browse/TIMOB-25539) on Android.
2. Verify that the app does not crash on startup.
3. Verify that a blue border is displayed around the ScrollableView.

---
**Test 2:**
Test that left/right page arrows are now density scaled. Note that the overlaid arrows used to be too tiny on high DPI devices.
1. Build and run the code attached to [TIMOB-25625](https://jira.appcelerator.org/browse/TIMOB-25625) on an "hdpi" and "xxhdpi" Android devices.
2. Scroll to the right in the ScrollableView on both devices.
3. Verify that the overlaid left/right page arrows are about the same size. (Arrows should no longer be tiny on a xxxhdpi device.)

---
**Test 3:**
Tests ScrollableView Ti.UI.SIZE support for [TIMOB-25626](https://jira.appcelerator.org/browse/TIMOB-25626). Also tests layout fix for [TIMOB-16878](https://jira.appcelerator.org/browse/TIMOB-16878).
1. Build and run the code attached to [TIMOB-25626](https://jira.appcelerator.org/browse/TIMOB-25626) on Android.
2. Note that the light gray container is the ScrollableView and the dark gray area is the Window.
3. Verify that view looks like [Android-good1.png](https://jira.appcelerator.org/secure/attachment/65127/Android-good1.png) on startup. Light gray container should not fill the window. (That's the old bad behavior.)
4. Scroll the light gray view to the right (by swiping left).
5. Verify view looks like [Android-good2.png](https://jira.appcelerator.org/secure/attachment/65123/Android-good2.png). Container should grow a bit bigger.
6. Scroll the light gray view to the right (by swiping left).
7. Verify view looks like [Android-good3.png](https://jira.appcelerator.org/secure/attachment/65126/Android-good3.png). container should grow a bit bigger.
8. Scroll the light gray view to the right (by swiping left).
9. Verify view looks like [Android-good4.png](https://jira.appcelerator.org/secure/attachment/65122/Android-good4.png). container should grow a bit bigger.
10. Scroll the light gray view to the right (by swiping left).
11. Purple view should now completely fill light gray container like [Android-good5.png](https://jira.appcelerator.org/secure/attachment/65125/Android-good5.png).

---
**Test 4:**
Test dynamically changing `cacheSize` property.
1. Build and run file `ScrollableViewCacheSizeTest.js` attached to [TIMOB-25634](https://jira.appcelerator.org/browse/TIMOB-25634) on Android.
2. Verify that the following gets logged...
```
[INFO] :   @@@ Initialized cacheSize: 5
[INFO] :   @@@ Property assigned cacheSize: 4
[INFO] :   @@@ Method assigned cacheSize: 3
```

---
**Test 5:**
1. Build and run file `ScrollableViewAddRemoveTest.js` attached to [TIMOB-15780](https://jira.appcelerator.org/browse/TIMOB-15780) on Android.
2. Note that this app automatically cycles through the pages programmatically via a timer.
2. Watch the Android log.
3. Verify that `@@@ [scrollend]` does not get logged when it programmatically changes pages. This is the fix for [TIMOB-15780](https://jira.appcelerator.org/browse/TIMOB-15780).
4. Scroll to another page with your finger. (Swipe left or right.)
5. Verify that a `[dragstart]`, `[scroll]`, `[dragend]`, and `[scrollend]` get logged when you scroll via the touchscreen yourself.
6. Tap the "Remove" button in the bottom-right corner.
7. An alert dialog is displayed. Tap the "Current" button to remove current page. This removes page by index and tests [TIMOB-5996](https://jira.appcelerator.org/browse/TIMOB-5996).
8. Repeat steps 6 and 7 until all 4 pages have been removed.

jquick-axway committed:
[TIMOB-16878] Android: Refactored ScrollableView

- [TIMOB-16878] Fixed bug where ScrollableView ignores top-most child view's layout properties.
  * Prevented properties width, height, top, bottom, left, and right from working.
  * Fixed by wrapping view with a TiCompositeLayout, which is needed to handle these layout properties.
- [TIMOB-25539] Fixed crash caused by adding non-Titanium view to ScrollableView.
- [TIMOB-25625] Modified left/right page arrows to be density scaled.
- [TIMOB-25626] Fixed ScrollableView to support Ti.UI.SIZE for width.
  * Android's "ViewPager" Java class does not natively support WRAP_CONTENT.
  * Refactored onMeasure() code to re-size width/height of container based on largest child view in this case.
- Removed API Level 9 guards from code. (Outdated code. Min API Level is currently 16.)

build commented:
<!--
  2 failure:  :disappointed_rel..., :microscope: Ther...
  0 warning: 
  1 messages
  
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/jquick-axway/titanium_mobile/blob/TIMOB-16878/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.1.0 test /Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9696-WEKQ5FB4SRSY2MDYWN6X7QK4FKWSCM2GAZECPZOTTQDDN7QFE6DQ
> grunt

Running "appcJs:src" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9696-WEKQ5FB4SRSY2MDYWN6X7QK4FKWSCM2GAZECPZOTTQDDN7QFE6DQ/dangerfile.js
   1:50  warning  'schedule' is defined but never used              no-unused-vars
  26:7   warning  'CURRENT_DIR' is assigned a value but never used  no-unused-vars

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9696-WEKQ5FB4SRSY2MDYWN6X7QK4FKWSCM2GAZECPZOTTQDDN7QFE6DQ/android/cli/commands/_build.js
  1636:9  warning  Blocks are nested too deeply (6)  max-depth

✖ 3 problems (0 errors, 3 warnings)


Running "continue:on" task

Running "retire:js" (retire) task
Ignoring {"paths":["growl"],"descriptors":[]}
Downloading https://raw.github.com/RetireJS/retire.js/master/repository/jsrepository.json ...
Checking: dangerfile.js
Checking: Gruntfile.js
Checking: apidoc/docgen.js
Checking: apidoc/lib/addon_generator.js
Checking: apidoc/lib/changes_generator.js
Checking: apidoc/lib/common.js
Checking: apidoc/lib/html_generator.js
Checking: apidoc/lib/jsca_generator.js
Checking: apidoc/lib/jsduck_generator.js
Checking: apidoc/lib/json_generator.js
Checking: apidoc/lib/modulehtml_generator.js
Checking: apidoc/lib/parity_generator.js
Checking: apidoc/lib/solr_generator.js
Checking: apidoc/validate.js
Checking: build/android.js
Checking: build/androidsdk.js
Checking: build/ant.js
Checking: build/docs.js
Checking: build/git.js
Checking: build/ios.js
Checking: build/packager.js
Checking: build/scons-build.js
Checking: build/scons-clean.js
Checking: build/scons-cleanbuild.js
Checking: build/scons-install.js
Checking: build/scons-package.js
Checking: build/scons-test.js
Checking: build/scons-update-node-deps.js
Checking: build/scons.js
Checking: build/titanium-sdk/lib/titanium.js
Checking: build/utils.js
Checking: build/windows.js
Checking: cli/commands/build.js
Checking: cli/commands/clean.js
Checking: cli/commands/create.js
Checking: cli/commands/project.js
Checking: cli/hooks/hyperloop-fix.js
Checking: cli/lib/creator.js
Checking: cli/lib/creators/app.js
Checking: cli/lib/creators/applewatch.js
Checking: cli/lib/creators/module.js
Checking: cli/lib/info.js
Checking: android/cli/commands/_build.js
Checking: android/cli/commands/_buildModule.js
Checking: android/cli/hooks/aar-transform.js
Checking: android/cli/hooks/package.js
Checking: android/cli/hooks/run.js
Checking: android/cli/lib/AndroidManifest.js
Checking: android/cli/lib/base-builder.js
Checking: android/cli/lib/detect.js
Checking: android/cli/lib/info.js
Checking: android/cli/tests/run.js
Checking: android/cli/tests/test-androidmanifest.js
Checking: iphone/cli/commands/_build.js
Checking: iphone/cli/commands/_buildModule.js
Checking: iphone/cli/hooks/frameworks.js
Checking: iphone/cli/hooks/hyperloop.js
Checking: iphone/cli/hooks/install.js
Checking: iphone/cli/hooks/package.js
Checking: iphone/cli/hooks/run.js
Checking: iphone/cli/lib/info.js
No vulnerabilities found.

Running "retire:node" (retire) task
Ignoring {"paths":["growl"],"descriptors":[]}
Downloading https://raw.github.com/RetireJS/retire.js/master/repository/npmrepository.json ...
Checking: .
Missing version for babel-eslint. Need to run npm install ?
Missing version for eslint-plugin-jsx-a11y. Need to run npm install ?
Missing version for eslint-plugin-react. Need to run npm install ?
Missing version for babel-eslint. Need to run npm install ?
Missing version for eslint-plugin-jsx-a11y. Need to run npm install ?
Missing version for eslint-plugin-react. Need to run npm install ?
No vulnerabilities found.

Running "continue:off" task

Running "ios_format:src" (ios_format) task

Running "android_format:src" (android_format) task
Fatal error: Error: Formatting incorrect on "android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIScrollableView.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='2578' length='5'></replacement>
<replacement offset='2786' length='5'></replacement>
<replacement offset='2806' length='1'>&#10;						   </replacement>
<replacement offset='10979' length='0'> </replacement>
<replacement offset='10987' length='0'> </replacement>
<replacement offset='11669' length='5'></replacement>
<replacement offset='11727' length='5'>&#10;														  </replacement>
<replacement offset='12252' length='5'></replacement>
<replacement offset='12310' length='5'>&#10;														   </replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

jquick-axway committed:
Android: Updated [TIMOB-16878] commit's code format.

jquick-axway committed:
Android/iOS/Windows: Added ScrollableView unit test "moveX-scrollTo". Replaces "moveNext" and "movePrev".

- Successfully tested on Android and iOS. (Have not tested on Windows.)

jquick-axway committed:
Android: Additional changes/refactoring of ScrollableView for [TIMOB-16878]

- [TIMOB-5996] Modified ScrollalbeView.removeView() to support integer indexes like iOS.
- [TIMOB-15780] Fixed ScrollableView.setCurrentPage() to not fire a "scrollend" event.
  * Event should only be fired when dragged or for methods that trigger scrolling animations.
- [TIMOB-25634] Modified ScrollableView so that "cacheSize" property can be set dynamically like iOS.
- [TIMOB-25636] Fixed bug where calling ScrollableView setters cause a crash if parent window is closed.
  * Bug was introduced in Titanium 7.0.0.

jquick-axway committed:
Merge branch 'master' into TIMOB-16878

jquick-axway committed:
Android: Updated code formatting for [TIMOB-16878].

garymathews committed:
Merge branch 'master' into TIMOB-16878

garymathews commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIScrollableView.java:
```diff
@@ -345,14 +414,41 @@ public void propertyChanged(String key, Object oldValue, Object newValue, KrollP
+	{
+		// Do not allow given size to be less than min. (iOS min size is 3.)
+		if (value < ScrollableViewProxy.MIN_CACHE_SIZE) {
+			StringBuilder stringBuilder = new StringBuilder();
```
Not sure the reason behind `StringBuilder` here? We could just do this:
```Java
Log.w(TAG, "ScrollableView 'cacheSize' cannot be set less than " +
            ScrollableViewProxy.MIN_CACHE_SIZE + ". Given value: " + value);
```

garymathews reviewed (approved):
CR: **PASS**

Although, a minor change could be made.

garymathews committed:
Merge branch 'master' into TIMOB-16878

garymathews commented:
@jquick-axway Looks like there's lint issues that need resolving

jquick-axway committed:
Merge branch 'master' into TIMOB-16878

jquick-axway committed:
Updated code format of ScrollableView unit test.

jquick-axway commented:
@sgtcoolguy, the new "ti.ui.scrollableview.test.js" removes tests "moveNext" and "movePrev" and replaces them with new unit test "moveX-scrollTo" which actually verifies that the view scrolled to the desired page.

Also, my "afterEach" change waits for the window to close before moving on to the next unit test, which based on my testing resolves the timeout issues that sometimes happens when attempting to open a window while the previous one is closing. I did the same with "MaskedImage" PR #10013. We shouldn't have to do this, but I don't have time to dig into why this is needed at the moment.

build commented:
<!--
  1 failure:  Tests have failed...
  0 warning: 
  1 messages
  1 markdown notices
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Tests have failed, see below for more information.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-9696/7/artifact/dist/mobilesdk-7.2.0.v20180502162646-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>

### Tests: 

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| ios.Titanium.Platform | netmask | 0 | file:///Users/ginsu/Library/Deve |
| ios.Titanium.Platform | address | 0 | file:///Users/ginsu/Library/Deve |

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

lokeshchdhry commented:
FR Passed.

1. [TIMOB-25626](https://jira.appcelerator.org/browse/TIMOB-25626) - `Ti.UI.SIZE` works as expected for scrollableView.
2. [TIMOB-25539](https://jira.appcelerator.org/browse/TIMOB-25539) - Adding a border to the scrollableView does not crash the app & works as expected.
3. [TIMOB-25625](https://jira.appcelerator.org/browse/TIMOB-25625) - The overlay arrows appear to be of the same size on `xhdpi`, `xxhdpi` & `xxxhdpi` devices.
4. [TIMOB-25634](https://jira.appcelerator.org/browse/TIMOB-25634) - Changing `cacheSize` dynamically works as expected.
5. [TIMOB-15780](https://jira.appcelerator.org/browse/TIMOB-15780) - `scrollend` does not fire when scrollableView is scrolled programmatically & does when scrolled manually.
6. [TIMOB-25636](https://jira.appcelerator.org/browse/TIMOB-25636) - No crash seen if parent window is closed.
7. [TIMOB-16878](https://jira.appcelerator.org/browse/TIMOB-16878) - The child view's layout properties are respected as expected.
8. [TIMOB-5996](https://jira.appcelerator.org/browse/TIMOB-5996) - Views can be removed successfully using integer indexes in `removeView`.

Studio Ver:     5.1.0.201804230827
SDK Ver:        7.2.0 local build
OS Ver:         10.13.4
Xcode Ver:      Xcode 9.3
Appc NPM:       4.2.13
Appc CLI:       7.0.3
Daemon Ver:     1.1.1
Ti CLI Ver:     5.1.0
Alloy Ver:      1.12.0
Node Ver:       8.9.1
NPM Ver:        5.5.1
Java Ver:       10
Devices:        ⇨ google Nexus 5 --- Android 6.0.1
                ⇨ google Nexus 6P --- Android 8.0.0
Emulator:       ⇨ Android 4.1.2

lokeshchdhry committed:
Merge branch 'master' into TIMOB-16878

lokeshchdhry committed:
Merge branch 'master' into TIMOB-16878

build commented:
<!--
  1 failure:  Tests have failed...
  0 warning: 
  1 messages
  1 markdown notices
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Tests have failed, see below for more information.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-9696/12/artifact/dist/mobilesdk-7.2.0.v20180508143408-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>

### Tests: 

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| android.Titanium.Network.HTTPClient | callbackTestForPOSTMethod | 60.049 | Error: timeout of 60000ms exceeded |
| android.Titanium.Network.HTTPClient | setCookieClearCookieWithMultipleHTTPClients | 60.009 | Error: timeout of 60000ms exceeded |
| android.Titanium.Network.HTTPClient | sendData | 60.006 | Error: timeout of 60000ms exceeded |
| android.Titanium.Network.HTTPClient | emptyPOSTSend | 60.014 | Error: timeout of 60000ms exceeded |
| android.Titanium.Network.HTTPClient | TIMOB-19042 | 60.011 | Error: timeout of 60000ms exceeded |
| android.Titanium.UI.ScrollableView | moveX-scrollTo | 20.013 | Error: timeout of 20000ms exceeded |
| android.Titanium.UI.TableView | appendSection and appendRow (TIMOB-25936) | 20.823 | Error: timeout of 5000ms exceeded |
| ios.Titanium.Network.HTTPClient | basic-auth failure | 60.007 | file:///Users/build/Librar |
| ios.Titanium.Network.HTTPClient | basic-auth success | 60.006 | file:///Users/build/Librar |
| ios.Titanium.Network.HTTPClient | setCookieClearCookieWithMultipleHTTPClients | 60.004 | file:///Users/build/Librar |
| ios.Titanium.Network.HTTPClient | sendData | 60.003 | file:///Users/build/Librar |
| ios.Titanium.Network.HTTPClient | largeFileWithRedirect | 60.006 | file:///Users/build/Librar |
| ios.Titanium.Network.HTTPClient | TIMOB-19042 | 60.007 | file:///Users/build/Librar |

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

lokeshchdhry committed:
Merge branch 'master' into TIMOB-16878

build commented:
<!--
  1 failure:  Tests have failed...
  0 warning: 
  1 messages
  1 markdown notices
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Tests have failed, see below for more information.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-9696/17/artifact/dist/mobilesdk-7.2.0.v20180509083108-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>

### Tests: 

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| android.Titanium.UI.ScrollableView | moveX-scrollTo | 20.029 | Error: timeout of 20000ms exceeded |
| android.Titanium.UI.TableView | appendSection and appendRow (TIMOB-25936) | 20.215 | Error: timeout of 5000ms exceeded |

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

sgtcoolguy committed:
Merge branch 'master' into TIMOB-16878
